ntoskrnl/ke/amd64/mp.c

```diff
@@ -0,0 +1,95 @@
+/*
+ * PROJECT:     ReactOS Kernel
+ * LICENSE:     GPL-2.0-or-later (https://spdx.org/licenses/GPL-2.0-or-later)
+ * PURPOSE:     Architecture source file to hold multiprocessor functions
+ * PROGRAMMERS:  Copyright 2021 Justin Miller <justinmiller100@gmail.com>
+ */
+
+/* INCLUDES *****************************************************************/
+
+#include <ntoskrnl.h>
+#define NDEBUG
+#include <debug.h>
+
+/* GLOBALS *******************************************************************/
+
+#define PHYSICAL_ADDRESS  LARGE_INTEGER
+PHYSICAL_ADDRESS HighestPhysicalAddress;
+ULONG ProcessorCount = 0;
+
+struct APInfo
+{
+    KPCR pcr;
+    KTSS tss;
+    ETHREAD thread;
+    KDESCRIPTOR GdtDesc;
+    KDESCRIPTOR IdtDesc;
+} APInfo;
+
+/* FUNCTIONS *****************************************************************/
+VOID
+NTAPI
+KeStartAllProcessors()
+{
+    BOOLEAN ApplicationProcessorsEnabled = FALSE;
+    HighestPhysicalAddress.QuadPart = UINT64_MAX;
+    KDESCRIPTOR GdtDesc, IdtDesc;
+    __sgdt(&GdtDesc.Limit);
+    __sidt(&IdtDesc.Limit);
+    do
+    {
+        ProcessorCount++;
+        /* Attempt to allocate memory for its new setup */
+        SIZE_T APInfo = sizeof(struct APInfo) + 2; 
+        PVOID PAPInfo = MmAllocateContiguousMemory(APInfo, HighestPhysicalAddress);
+
+        if (!PAPInfo)
+        {
+            ASSERT("KeStartAllProcessors: Memory Allocation has failed");
+        }
+
+        RtlCopyMemory((PVOID)((ULONG_PTR)PAPInfo + sizeof(KPCR) + sizeof(KTSS)), (PVOID)GdtDesc.Base, GdtDesc.Limit + 1);
```

Comment:
There is a structure definition for it, so use that.

Reply:
And what is that structure? I don't see any structure like:
```c
typedef struct _some_struct
{
    KPCR Pcrb;
    KTSS Tss;
#ifdef _M_AMD64
    // Here: DoubleFaultStack
    // Here: MachineCheckStack
    // Here: NMIStack
#endif
    ETHREAD InitialThread;
#ifdef _M_IX86
    KTSS DoubleFaultTSS;
    KTSS NmiTSS;
    // Here: ExceptionStack;
#endif
    // Here: The full GDT for the given processor.
    // Here: The full IDT for the given processor.
} my_struct;
```
which is the one necessary for holding the full per-CPU data information, as explained in https://www.geoffchappell.com/studies/windows/km/ntoskrnl/inc/ntos/kpcr.htm .

Reply:
1. My comment has nothing to do with arcitecture specific fields.
2. There is no problem with having a architecture specific definition of that structure.
3. You wouldn't need any of that, because you can put all of that into the processor state.

Reply:
The `PROCESSOR_STATE` structure contains "pointers" to the actual data that needs to be stored somewhere (the stacks, the ETHREAD, the KPCR , for that given processor), and this is that actual data that needs to be allocated and initialized.

Reply:
I was primarily referring to the use of pointer arithmetic. `There is a structure definition for it` means there is one in line 20 of this file, so there it should be used instead of `(ULONG_PTR)PAPInfo + sizeof(KPCR) + ...`
The comment about using the PROCESSOR_STATE was related to copying the KDESCRIPTOR, which is part of KPROCESSOR_STATE. But what is done here is even more crazy, as it copies the entire GDT into a KDESCPRIPTOR field.